### DISCUSSION TITLE
How to check asynchronously user status in `beforeLoad`?

### DISCUSSION THREAD

**Discussion by pooryamoh** [TanStack/router#1668] [Q&A]
Hi,

I have tried to follow the example of a protected route with modification for checking the user status from the server asynchronously, but I wasn't successful because `beforeLoad` will return before the context state changes. How can I fix this problem?

auth.tsx:
```javascript
import axios from "axios";
import {
  createContext,
  ReactNode,
  useState,
  useCallback,
  useEffect,
} from "react";

export interface IAuthContext {
  isAuthenticated: boolean;
  login: (username: string, password: string) => void;
  logout: () => void;
  user: IUser | null;
}
export interface IUser {
  id: string;
  firstName: string;
  lastName: string;
  username: string;
  jwtToken: string;
}

export const AuthContext = createContext<IAuthContext>({} as IAuthContext);

export function AuthProvider({ children }: { children: ReactNode }) {
  const [user, setUser] = useState<IUser | null>(null);
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(false);

  useEffect(() => {
    const isLogedIn = async () => {
      try {
        const response = await axios.post<IUser>(
          "https://localhost:7271/users/refresh-token",
          {},
          {
            withCredentials: true,
          }
        );
        if (response.status === 200) {
          setUser(() => response.data);
          setIsAuthenticated(() => true);
        }
      } catch (error) {
        console.error(error)
      }
    };
    isLogedIn();
  }, []);

  const logout = useCallback(() => {}, []);

  const login = useCallback(async (username: string, password: string) => {
    try {
      const response = await axios.post<IUser>(
        "https://localhost:7271/users/authenticate",
        {
          username: username,
          password: password,
        },
        {
          withCredentials: true,
        }
      );
      if (response.status === 200) {
        setUser(() => response.data);
        setIsAuthenticated(() => true);
      }
    } catch (error) {
      console.error(error);
    }
  }, []);

  return (
    <AuthContext.Provider value={{ isAuthenticated, user, login, logout }}>
      {children}
    </AuthContext.Provider>
  );
}
```

index.tsx:
```javascript
import { createFileRoute, redirect } from "@tanstack/react-router";

export const Route = createFileRoute("/")({
  beforeLoad: ({ context, location }) => {
    if (!context.auth.isAuthenticated) {
      throw redirect({
        to: "/login",
        search: {
          redirect: location.href,
        },
      });
    } else {
      throw redirect({
        to: "/dashboard/",
      });
    }
  },
});

```

**Comment by SzelamC** [comment]
Facing the same issue, i also see the authenticated route example provided in [documentation](https://tanstack.com/router/latest/docs/framework/react/examples/authenticated-routes) ,

login.tsx
```typescript
import * as React from 'react'
import {
  createFileRoute,
  redirect,
  useRouter,
  useRouterState,
} from '@tanstack/react-router'
import { z } from 'zod'

import { useAuth } from '../auth'
import { sleep } from '../utils'

const fallback = '/dashboard' as const

export const Route = createFileRoute('/login')({
  validateSearch: z.object({
    redirect: z.string().optional().catch(''),
  }),
  beforeLoad: ({ context, search }) => {
    if (context.auth.isAuthenticated) {
      throw redirect({ to: search.redirect || fallback })
    }
  },
  component: LoginComponent,
})

function LoginComponent() {
  const auth = useAuth()
  const router = useRouter()
  const isLoading = useRouterState({ select: (s) => s.isLoading })
  const navigate = Route.useNavigate()
  const [isSubmitting, setIsSubmitting] = React.useState(false)

  const search = Route.useSearch()

  const onFormSubmit = async (evt: React.FormEvent<HTMLFormElement>) => {
    setIsSubmitting(true)
    try {
      evt.preventDefault()
      const data = new FormData(evt.currentTarget)
      const fieldValue = data.get('username')

      if (!fieldValue) return
      const username = fieldValue.toString()
      await auth.login(username)

      await router.invalidate()

      // This is just a hack being used to wait for the auth state to update
      // in a real app, you'd want to use a more robust solution
      await sleep(1)

      await navigate({ to: search.redirect || fallback })
    } catch (error) {
      console.error('Error logging in: ', error)
    } finally {
      setIsSubmitting(false)
    }
  }

  const isLoggingIn = isLoading || isSubmitting

  return (
    <div className="p-2 grid gap-2 place-items-center">
      <h3 className="text-xl">Login page</h3>
      {search.redirect ? (
        <p className="text-red-500">You need to login to access this page.</p>
      ) : (
        <p>Login to see all the cool content in here.</p>
      )}
      <form className="mt-4 max-w-lg" onSubmit={onFormSubmit}>
        <fieldset disabled={isLoggingIn} className="w-full grid gap-2">
          <div className="grid gap-2 items-center min-w-[300px]">
            <label htmlFor="username-input" className="text-sm font-medium">
              Username
            </label>
            <input
              id="username-input"
              name="username"
              placeholder="Enter your name"
              type="text"
              className="border border-gray-300 rounded-md p-2 w-full"
              required
            />
          </div>
          <button
            type="submit"
            className="bg-blue-500 text-white py-2 px-4 rounded-md w-full disabled:bg-gray-300 disabled:text-gray-500"
          >
            {isLoggingIn ? 'Loading...' : 'Login'}
          </button>
        </fieldset>
      </form>
    </div>
  )
}
```
There is a comment mention that need to implement a robust way for waiting the state update, but i am not sure how to do that and how it consider a robust way.

I try below and not work
```typescript
  useEffect(() => {
    if (isAuthenticated) {
      navigate({ to: "/somepath" });
    }
  }, [isAuthenticated]);
```

**Reply by pooryamoh** [reply]
The solution I found was using Zustand instead of router context. Beacuse I can access Zustand  without hook.

**Reply by SzelamC** [reply]
I try flushSync and also solved the issue, but i am not sure isn't flushSync a 'robust' way to deal with this problem. Since it is frustrated that zustand is a must dependency to solve this issue.

**Reply by KevinGuedes** [reply]
> I try flushSync and also solved the issue, but i am not sure isn't flushSync a 'robust' way to deal with this problem. Since it is frustrated that zustand is a must dependency to solve this issue.

Could you share your solution with `flushSync` please?

**Reply by SzelamC** [reply]
This is a simple solution which you may want to change the logic. 
Basically, this happen because on `_auth.tsx` beforeLoad context value isn't update when user login and `setUser(user)` not execute synchronously causing beforeLoad context value is stale.
`flushSync` forcing react to update the state synchronously.

`_auth.tsx`
```typescript
export const Route = createFileRoute("/_auth")({
  beforeLoad: ({ context }) => {
    if (!context.auth.isAuthenticated) {
      throw redirect({
        to: "/login",
      });
    }
  },
  component: AuthLayout,
});
```

`login.tsx`
```typescript
export function AuthProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState<User | null>(getUserFromLocalStorage());
  const isAuthenticated = !!user;

  useEffect(() => {
    const user = getUserFromLocalStorage();
    if (user) {
      setUser(user);
    }
  }, []);

  const login = useCallback(async (data: LoginSchema) => {
    const user = await fakeApi.login(data);
    if (user) {
      localStorage.setItem(key, JSON.stringify(user));
      flushSync(() => {
        setUser(user);
      });
    }
  }, []);

  return (
    <AuthContext.Provider value={{ isAuthenticated, user, login }}>
      {children}
    </AuthContext.Provider>
  );
}

```

**Reply by KevinGuedes** [reply]
@SzelamC Thank you!

**Reply by surious-type** [reply]
@SzelamC You're my hero, all the best!

**Reply by ParasSolanki** [reply]
If you don't want to include other dependencies like [Zustand](https://zustand.docs.pmnd.rs/) because you can access store data outside of react (Love Zustand btw). So, you can use [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore) hook which will lets you create a external store and subscribe to its changes. 

For the authenticated state you can create a external store `authState` like this which will have following properties:
1. `isAuthenticated` indicate whether user is authenticated or not.
2. `isInitialLoading` defaults to true and whenever the first fetch for a auth query.
3. `user` authenticated user.

```tsx
// src/hooks/use-auth.ts

import { useQuery } from "@tanstack/react-query";
import { meQuries } from "~/common/keys/me";
import type { AuthMe } from "~/common/keys/me";
import { HTTPError } from "ky";
import { useSyncExternalStore } from "react";

type SubscribeListener = () => void;

let listeners = [] as Array<SubscribeListener>;
export type AuthState = {
  isAuthenticated: boolean;
  isInitialLoading: boolean;
  user: AuthMe | undefined;
};

export const authState = {
  isAuthenticated: false,
  isInitialLoading: true,
  user: undefined,
} as AuthState;

export const authStore = {
  setIsInitialLoading(isInitialLoading: AuthState["isInitialLoading"]) {
    authState.isInitialLoading = isInitialLoading;
    emitChange();
  },
  setIsAuthenticated(isAuthenticated: AuthState["isAuthenticated"]) {
    authState.isAuthenticated = isAuthenticated;
    emitChange();
  },
  setUser(user: AuthState["user"]) {
    authState.user = user;
    emitChange();
  },
  subscribe(listener: SubscribeListener) {
    listeners = [...listeners, listener];
    return () => {
      listeners = listeners.filter((l) => l !== listener);
    };
  },
  getSnapshot() {
    return authState;
  },
};

function emitChange() {
  for (const listener of listeners) {
    listener();
  }
}

export function useAuthState() {
  return useSyncExternalStore(authStore.subscribe, authStore.getSnapshot);
}

export function useSession() {
  const authState = useAuthState();
  const session = useQuery(meQuries.session());

  if (session.data) {
    authStore.setIsAuthenticated(true);
    authStore.setUser(session.data.data.user);
  }

  // unauthorized
  if (
    session.error &&
    authState.isAuthenticated &&
    session.error instanceof HTTPError
  ) {
    authStore.setIsAuthenticated(false);
    authStore.setUser(undefined);
    window.location.href = "/signin";
  }

  if (authState.isInitialLoading) {
    authStore.setIsInitialLoading(session.isLoading);
  }

  if (session.isError) {
    authStore.setIsInitialLoading(false);
  }

  return session;
}
```

Add the authstate store in router context. You can also use [TanstackQuery](https://tanstack.com/query/latest) which is awesome for managing async state. Add the queryClient to router context as well.

```tsx
// src/routes/__root.tsx

import { QueryClient } from "@tanstack/react-query";
import { createRootRouteWithContext, Outlet } from "@tanstack/react-router";
import { env } from "~/env";
import { AuthState } from "~/hooks/use-auth";
import * as React from "react";

const TanStackRouterDevtools = env.PROD
  ? () => null // Render nothing in production
  : React.lazy(() =>
      import("@tanstack/router-devtools").then((res) => ({
        default: res.TanStackRouterDevtools,
      })),
    );

type RouterContext = {
  queryClient: QueryClient;
  authState: AuthState;
};

export const Route = createRootRouteWithContext<RouterContext>()({
  component: Root,
});

function Root() {
  return (
    <>
      <Outlet />
      <React.Suspense>
        <TanStackRouterDevtools position="bottom-left" />
      </React.Suspense>
    </>
  );
}
```

In the App.tsx before initializing the router we will check for the authenticated flag and if query is loading show loading spinner.

```tsx
// src/App.tsx

import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { ReactQueryDevtools } from "@tanstack/react-query-devtools";
import { createRouter, RouterProvider } from "@tanstack/react-router";
import { Toaster } from "~/components/ui/sonner";
import { Loader2Icon } from "lucide-react";
import { authState, useAuthState, useSession } from "./hooks/use-auth";
// Import the generated route tree
import { routeTree } from "./routeTree.gen";

const queryClient = new QueryClient();

// Create a new router instance
const router = createRouter({
  routeTree,
  context: {
    queryClient,
    authState,
  },
  defaultPreload: "intent",
});


// Register the router instance for type safety
declare module "@tanstack/react-router" {
  interface Register {
    router: typeof router;
  }
}

export function App() {
  return (
    <QueryClientProvider client={queryClient}>
      <AppRoutes />
      <ReactQueryDevtools position="right" />
      <Toaster richColors className="font-sans" position="bottom-right" />
    </QueryClientProvider>
  );
}

function AppRoutes() {
  useSession();
  const authState = useAuthState();

  if (authState.isInitialLoading) {
    return (
      <div className="flex h-screen w-full items-center justify-center p-4">
        <Loader2Icon className="size-10 animate-spin text-foreground" />
      </div>
    );
  }

  return (
    <RouterProvider router={router} context={{ queryClient, authState }} />
  );
}
```

When you signed in the user update the authstate store values as well. So, that after signin you can redirect to authenticated routes.

```tsx
// src/routes/signin/route.lazy.tsx

function SignInForm() {
  const navigate = useNavigate();
  const queryClient = useQueryClient();
  const form = useForm({
    defaultValues: {
      email: "",
      password: "",
    },
  });

  const { isPending, mutate } = useMutation({
    mutationKey: ["signin"],
    mutationFn: async (values: AuthValues) => {
      // handle user signin api
    },
    onSuccess: () => {
      // on success
    },
    onError: async (error) => {
      // on error
    },
  });

  function onSubmit(values: AuthValues) {
    mutate(values, {
      onSuccess: () => {
        authStore.setIsAuthenticated(true);
        queryClient.invalidateQueries({ queryKey: meKeys.all });
        navigate({
          to: "/",
        });
      },
    });
  }

  return (
    <Form {...form}>
      <form className="mt-6" onSubmit={form.handleSubmit(onSubmit)}>
        ...
      </form>
    </Form>
  );
}
```

For the authenticated routes create a pathless layout route in routes `routes/_auth/route.tsx` which will check for authenticated flag and if user is not authenticated it will redirect them to signin page.
Also since useSession() is called in App.tsx and we are using tanstack query it will do window focus background fetches for session query which we use to reload the route if user session expires it will redirect it to signin page. For that router provide `shouldReload()` method.

```tsx
import { createFileRoute, redirect } from "@tanstack/react-router";

export const Route = createFileRoute("/_auth")({
  beforeLoad: ({ context }) => {
    if (!context.authState.isAuthenticated) {
      throw redirect({
        to: "/signin",
        search: {
          redirectUrl: location.href,
        },
      });
    }
  },
  shouldReload({ context }) {
    return !context.authState.isAuthenticated; 
  },
});
```
If you want the full code you can checkout this repo: https://github.com/ParasSolanki/roles-permissions

**Reply by leomelzer** [reply]
Thank you @ParasSolanki, I've adpated this for usage with supabase:

```ts
import type { AuthUser, Session } from "@supabase/supabase-js";
import { useEffect, useState, useSyncExternalStore } from "react";
import { supabase } from "./supabase";

type SubscribeListener = () => void;

let listeners = [] as Array<SubscribeListener>;

export type AuthState = {
  isAuthenticated: boolean;
  isInitialLoading: boolean;
  user: AuthUser | undefined;
};

export const authState = {
  isAuthenticated: false,
  isInitialLoading: true,
  user: undefined,
} as AuthState;

export const authStore = {
  setIsInitialLoading(isInitialLoading: AuthState["isInitialLoading"]) {
    authState.isInitialLoading = isInitialLoading;
    emitChange();
  },
  setIsAuthenticated(isAuthenticated: AuthState["isAuthenticated"]) {
    authState.isAuthenticated = isAuthenticated;
    emitChange();
  },
  setUser(user: AuthState["user"]) {
    authState.user = user;
    emitChange();
  },
  subscribe(listener: SubscribeListener) {
    listeners = [...listeners, listener];
    return () => {
      listeners = listeners.filter((l) => l !== listener);
    };
  },
  getSnapshot() {
    return authState;
  },
};

function emitChange() {
  for (const listener of listeners) {
    listener();
  }
}

export function useAuthState() {
  return useSyncExternalStore(authStore.subscribe, authStore.getSnapshot);
}

export function useSession() {
  const authState = useAuthState();
  const [isLoading, setIsLoading] = useState(true);
  let session: Session | null = null;

  useEffect(() => {
    supabase.auth
      .getSession()
      .then(({ data: { session: currentSession }, error }) => {
        if (error) {
          authStore.setIsAuthenticated(false);
          authStore.setIsInitialLoading(false);
          authStore.setUser(undefined);
          window.location.href = "/login";
        } else {
          session = currentSession;
          if (currentSession?.user) {
            authStore.setIsAuthenticated(true);
            authStore.setUser(currentSession.user);
          }
          setIsLoading(false);
        }
      });

    const { data } = supabase.auth.onAuthStateChange((event, session) => {
      if (
        session &&
        (event === "PASSWORD_RECOVERY" ||
          // Detect if the user doesn't have a password yet (e.g. when coming from an invite link)
          !session.user.user_metadata.hasPassword)
      ) {
        window.location.href = "/change-password";
      }

      if (session) {
        authStore.setIsAuthenticated(true);
        authStore.setUser(session.user);
      }
    });

    return () => data.subscription.unsubscribe();
  }, []);

  if (authState.isInitialLoading) {
    authStore.setIsInitialLoading(isLoading);
  }

  return session;
}

```

This works nicely and handles all edge cases I can think of right now nicely.

**Comment by VladKopylov** [comment]
There was also a problem with async user status checking. If possible, please give more examples on how to deal with such cases

**Comment by KevinGuedes** [comment]
I would love to see this 'more robust' solution without having to add any other dependencies haha, facing the same problem here

**Comment by maxwell-oroark** [comment]
I'm facing the same issue as well. Even when a firebase user is logged in it's momentarily async to load the user into state and by then the before load has already fired and redirected user to `/login` page

**Reply by tigawanna** [reply]
try fetching the user and storing the value in the router context  Instead of doing the fetch in beforeLoad

I did this by fetching just before the router provider bit inside the query provider 

example:
https://github.com/tigawanna/logic-coding-challenge

**Reply by maxwell-oroark** [reply]
thanks for the reply @tigawanna - I took a look and found this line I think you're alluding to: https://github.com/tigawanna/logic-coding-challenge/blob/bb539302ab1e892170886a6d9bac6710b8eb571a/src/main.tsx#L56

here's my issue, my firebase user momentarily resolves to `null`. Let's say the user loads the app on an authenticated route. That null value propagate through the context and hits the beforeLoad check, beforeLoad check boots them to `/login` page. Right after that the firebase onAuthStateChange() callback fires and populates the user. Now the user is on `/login` page even though the are signed in. 

I've also tried the non-redirect approach. The downside of this approach is login screen flash. User loads the app on authenticated route, user not defined yet, flash login page instead of <Outlet /> component, then user becomes available and the screen flashes back to the app. Both seem not ideal...

**Reply by tigawanna** [reply]
have you tried fetching the current user at the top most part of the app before you hit any pages and rely less onAuthChanged i had an example where i got it working with firebase. use query to cache it for a like an hour  so that it'll only be null if you manually invalidate that query

**Reply by maxwell-oroark** [reply]
Yeah so you're suggesting basically wrap the onAuthChanged in a query hook of some sort that has a notion of fetched and loading states, then manage the loading, resolved, error states _outside_ of tanstack router with some top level JSX conditionals before rendering the router?

**Reply by tigawanna** [reply]
firebase has a get current user method it's more deterministic

**Reply by maxwell-oroark** [reply]
Yeah, I guess I got excited about tanstack's promise of being able to declare all your data dependencies up front and then get loading, error states out of the box.  A little disappointed that the first data dependency I have to implement ( active user ) I have to re-invent the wheel a bit with conditional logic outside of the router.

**Reply by maxwell-oroark** [reply]
On second thought... I could just wrap the observable in a promise and await in `beforeLoad`... 

```
  beforeLoad: async () => {
    const user = await getCurrentUserAsync();
    if (!user) {
      throw redirect({
        to: "/login",
        search: {
          redirect: location.href,
        },
      });
    }
  },
```

**Reply by tigawanna** [reply]
there's an explanation somewhere in the discord to that behavior but it's the most  ergonomic route guard implementation I've used

**Comment by FredricW** [comment]
I chose to wrap my Auth instance in a promise that resolves when it is done loading:

```tsx
interface RouterContext {
  auth: Promise<AuthInterface>,
  ...
}
```
```tsx
// main.tsx

...

let resolveAuthClient: (client: AuthInterface) => void;
export const authClient: Promise<AuthInterface> = new Promise(
  (resolve) => { resolveAuthClient = resolve }
);

function InnerApp() {
  const auth = useAuth();

  useEffect(() => {
    if (auth.isLoading) return;

    resolveAuthClient(auth);
  }, [auth, auth.isLoading]);

  return <RouterProvider router={router} context={{ auth: authClient }} />;
}
```
```tsx
// _auth.tsx
export const Route = createFileRoute('/_auth')({
  pendingComponent: Spinner,
  async beforeLoad({ context }) {
    const auth = await context.auth; // Wait for auth to be done loading
    // Do your stuff here
  },
  component: () => <Outlet />,
});
```

Then any route under the folder `src/routes/_auth/` will trigger auth.

EDIT: Cutting edge ES2024 edition:
```tsx
const authClient = Promise.withResolvers<AuthInterface>()
function InnerApp() {
  const auth = useAuth();

  useEffect(() => {
    if (auth.isLoading) return;

    authClient.resolve(auth);
  }, [auth, auth.isLoading]);

  return <RouterProvider router={router} context={{ auth: authClient }} />;
}
```
EDIT2: Removed `new` keyword in the example as suggested by @isaackoz

**Reply by dnlsandiego** [reply]
Great tip on the `Promise.withResolvers` use. But what's the reason for defining `authClient` at the module level/scope instead of inside `InnerApp` as state or stored in a ref?

**Reply by isaackoz** [reply]
Should be `const authClient = Promise.withResolvers<AuthInterface>()` (remove `new`).

 Promise.withResolvers already returns a promise instance.

**Reply by FredricW** [reply]
@dnlsandiego sorry for the late answer. It is a little bit off topic, but we actually have it in a different file and export it. The reason we do it is because we have a middleware for our api clients that adds `Authentication` headers automatically using the same technique. They simply import the `authClient` and await it before getting and setting the token.

**Reply by hieunguyent12** [reply]
Hi @FredricW, this is a great solution. When it comes to the initial page load when a new user visits the page on a slow connection, how do you handle that? Are you using pendingMs and pendingComponent to show some kind of loading state while the user waits? If that is the case then what happens if the user's connection is very slow and the loading state is shown for a prolonged period which I think would not be ideal because they have to wait for the login page to be shown.

**Reply by FredricW** [reply]
@hieunguyent12 we set a defaultPendingComponent when configuring the router which will get rendered when beforeLoad (or any other loader) is still pending. I didn't see pendingMs to be necessary because we have to load more data after authentication which is quite slow so the spinner will appear for some time regardless of how fast the auth is anyway.

**Reply by maxwell-oroark** [reply]
For anyone trying this solution, I actually had to pass in the `.promise` because:
```
Promise.withResolvers() // returns { promise, resolve, reject } 
```

```
const authClient = Promise.withResolvers<AuthInterface>()
function InnerApp() {
  const auth = useAuth();

  useEffect(() => {
    if (auth.isLoading) return;

    authClient.resolve(auth);
  }, [auth, auth.isLoading]);

  return <RouterProvider router={router} context={{ auth: authClient.promise }} />;
}
```

**Reply by iamsahebgiri** [reply]
How will you handle logout? I signed out but the promise in the context is already resolved. 

I tried `router.invalidate()`, that does re-run the `beforeLoad()`, but I want to make the auth to be `null` after logout.

**Reply by jo3-l** [reply]
@iamsahebgiri I ran into the same problem with this code. Instead of storing the promise directly in the router context you can provide a method `loadAuth` that only returns the promise if authentication is still loading, and otherwise returns the latest authentication context. I wrote up the details in my comment at https://github.com/TanStack/router/discussions/4320#discussioncomment-13982671.

**Reply by BranislavLazic** [reply]
Please keep in mind that this isn't a complete solution and it works well only for a very static `AuthInterface` context. In the case if the context of `AuthInterface` changes, then those changes will not be reflected in the routing (old value for `context.auth` call in `beforeLoad` will be still present). E.g. I encountered this issue while switching between Clerk organizations in my app, which requires `AuthInterface` context to reload its state. 

Since this switching is not a very common action in my app, I went for a simple solution where an element that triggers the switching is calling `navigate({ to: "/", replaceDocument: true })`. This will force the context to reload.

**Comment by kad5** [comment]
I used router.invalidate() inside my beforeLoad after updating the context state and it worked

**Comment by TmLev** [comment]
If you're using TanStack Query to manage auth state, you can simply do this:

```tsx
queryClient.resetQueries({ queryKey: authQueryKey });
```

just before the `router.invalidate()` call

**Reply by walidsj** [reply]
Thank you. This is just work! Use `queryClient.resetQueries({ queryKey: authQueryKey });` Instead of `queryClient.invalidateQueries({ queryKey: authQueryKey });`

**Reply by LiteBluSky** [reply]
This right here, was the answer to a fight that has lasted a week, and here it is, burried in the deepest of depths of github. Thank you!
I would like to understand the difference of these 2 solutions beeing the following: 

```javascript
queryClient.resetQueries({ queryKey: ['key'] })
queryClient.invalidateQueries({ queryKey: ['key'] })
```

**Reply by TmLev** [reply]
`resetQueries` resets query's state to initial (empty or `initialData`).
`invalidateQueries` performs a refetch.

Docs:
* https://tanstack.com/query/v5/docs/reference/QueryClient#queryclientresetqueries
* https://tanstack.com/query/v5/docs/reference/QueryClient#queryclientinvalidatequeries

**Comment by alexander-zuev** [comment]
This thread is a lifesaver! Tried to implement what @FredricW originally proposed, but I ran into an issue where the auth promise was not re-created which led to an incorrect sign out flow - user signed out but the router never fired again and re-directed.

Here is the solution I came up with to make Supabase Auth work with Tanstack router.

Instead of passing a static promise that can become "stale", pass a **function** that returns a fresh promise each time it's called.
### Key parts
**- Define the context interface:`__root.tsx`**
``` tsx
interface RouterContext {
  queryClient: QueryClient;
  waitForAuth: () => Promise<{ user: User | null }>; // Function, not promise
}

export const Route = createRootRouteWithContext<RouterContext>()({
  component: RootComponent,
  notFoundComponent: NotFoundComponent,
});
```
**- Create the helper function:`main.tsx`**
``` tsx
// Helper that returns a fresh promise each time
export function waitForAuth(): Promise<{ user: User | null }> {
  const { user, isInitialized } = useAuthStore.getState();

  if (isInitialized) {
    return Promise.resolve({ user });
  }

  return new Promise(resolve => {
    const unsub = useAuthStore.subscribe(state => {
      if (state.isInitialized) {
        unsub();
        resolve({ user: state.user });
      }
    });
  });
}

const router = createRouter({
  routeTree,
  context: {
    queryClient,
    waitForAuth, // Pass the function
  },
});
```
**`_auth.tsx` - Use in route guards:**
``` tsx
export const Route = createFileRoute('/_auth')({
  beforeLoad: async ({ context }) => {
    const { user } = await context.waitForAuth(); // Call the function
    
    if (!user) {
      throw redirect({ to: '/' });
    }
  },
  component: () => <Outlet />,
});
```
### Why This Works
1. **Fresh promises**: Each call creates a new promise with current state `waitForAuth()`
2. **No stale promises**: Solves the "dead promise" issue after sign-out
3. **Immediate resolution**: If auth is ready, resolves instantly
4. **Waits when needed**: If auth is loading, waits for initialization

P.S. Written by Sonnet 4 from the solution code

**Comment by jo3-l** [comment]
I wrote up my explanation of the problem and a solution in https://github.com/TanStack/router/discussions/4320#discussioncomment-13982671. It's similar in principle to some of the above ideas that use deferred promises (https://github.com/TanStack/router/discussions/1668#discussioncomment-10634735 and https://github.com/TanStack/router/discussions/1668#discussioncomment-13485090). The differences are that my idea

- avoids a subtle issue in the first solution where the auth context can be stale, as noted in the second solution,
- and is a little more general than the second solution (doesn't require the authentication provider to expose an API to subscribe to changes directly)
